Sources/Auth/Apple/AppleAuthProvider.swift

```diff
@@ -0,0 +1,173 @@
+            let identityTokenString = String(data: identityToken, encoding: .utf8) else {
+        delegate?.appleAuthProviderDidFail(with: .invalidIdentityToken)
+        return
+      }
+      
+      // store userId for later
+      storage.set(credential.user, forKey: userIdStorageKey)
+      
+      let displayName = [credential.fullName?.givenName, credential.fullName?.familyName]
+        .compactMap { $0 }
+        .joined(separator: " ")
+      
+      delegate?.appleAuthProviderDidSignIn(with: .init(token: identityTokenString,
+                                                       name: displayName,
+                                                       email: credential.email))
+    case _:
+      delegate?.appleAuthProviderDidFail(with: .unhandledAuthorization)
+    }
+  }
+  
+  public func authorizationController(controller: ASAuthorizationController, didCompleteWithError error: Swift.Error) {
+    delegate?.appleAuthProviderDidFail(with: .system(error))
+  }
+}
+
+// MARK: - ASAuthorizationControllerPresentationContextProviding
+extension AppleAuthProvider: ASAuthorizationControllerPresentationContextProviding {
+  public func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
+    presentationAnchor ?? UIWindow()
+  }
+}
+
+// MARK: - Private Methods
+private extension AppleAuthProvider {
+  func appleSignIn(with nonce: Nonce?) {
+    let request = authProvider.createRequest()
+    request.requestedScopes = [.fullName, .email]
+    
+    switch nonce {
+    case .random(let length):
+      guard length > 0 else {
+        delegate?.appleAuthProviderDidFail(with: .invalidNonceLength)
+        return
+      }
+      request.nonce = generateRandomNonceString(length: length).sha256
+    case .none:
+      break
+    }
+    
+    let controller = ASAuthorizationController(authorizationRequests: [request])
+    controller.delegate = self
+    controller.presentationContextProvider = self
+    controller.performRequests()
+  }
+  
+  func generateRandomNonceString(length: UInt = 32) -> String {
+    let charset: [Character] = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")
+    let result = (0..<length).compactMap { _ in charset.randomElement() }
+    guard result.count == length else { fatalError("Unable to generate nonce!") }
+    return String(result)
+  }
+  
+  func setupCredentialsRevokeListener() {
+    NotificationCenter.default.addObserver(self,
+                                           selector: #selector(appleCredentialRevoked),
+                                           name: ASAuthorizationAppleIDProvider.credentialRevokedNotification,
+                                           object: nil)
+  }
+}
+
+// MARK: - Actions
+private extension AppleAuthProvider {
+  @objc func appleCredentialRevoked() {
+    delegate?.appleAuthProviderIsAuthorized(false)
+  }
+}
+
+// MARK: - Private String Extension Methods
+private extension String {
```

Comment:
Maybe we can move this to *Auth Core* module or to the PovioKit *Extensions* since it is not directly related only to the AppleAuth?

Reply:
I wanted to make Auth modules independent from any other. Good point though. I'll think about it.